crates/turborepo-ui/src/wui/mod.rs

```diff
@@ -0,0 +1,275 @@
+}
+
+#[derive(Debug, Clone)]
+pub struct WebUISender {
+    pub tx: tokio::sync::broadcast::Sender<WebUIEvent>,
+}
+
+impl WebUISender {
+    pub fn start_task(&self, task: String, output_logs: OutputLogs) {
+        self.tx
+            .send(WebUIEvent::StartTask { task, output_logs })
+            .ok();
+    }
+
+    pub fn end_task(&self, task: String, result: TaskResult) {
+        self.tx.send(WebUIEvent::EndTask { task, result }).ok();
+    }
+
+    pub fn status(&self, task: String, status: String, result: CacheResult) {
+        self.tx
+            .send(WebUIEvent::Status {
+                task,
+                status,
+                result,
+            })
+            .ok();
+    }
+
+    pub fn set_stdin(&self, _: String, _: Box<dyn Write + Send>) {
+        warn!("stdin is not supported (yet) in web ui");
+    }
+
+    pub fn task(&self, task: String) -> TaskSender {
+        TaskSender {
+            name: task,
+            handle: UISender::Wui(self.clone()),
+            logs: Default::default(),
+        }
+    }
+
+    pub fn stop(&self) {
+        self.tx.send(WebUIEvent::Stop).ok();
+    }
+
+    pub fn update_tasks(&self, tasks: Vec<String>) -> Result<(), crate::Error> {
+        self.tx
+            .send(WebUIEvent::UpdateTasks { tasks })
+            .map_err(Error::Broadcast)?;
+
+        Ok(())
+    }
+
+    pub fn output(&self, task: String, output: Vec<u8>) -> Result<(), crate::Error> {
+        self.tx
+            .send(WebUIEvent::TaskOutput { task, output })
+            .map_err(Error::Broadcast)?;
+
+        Ok(())
+    }
+}
+
+// Specific events that the websocket server can send to the client,
+// not all the `Event` types from the TUI
+#[derive(Debug, Clone, Serialize)]
+#[serde(tag = "type", content = "payload")]
+pub enum WebUIEvent {
+    StartTask {
+        task: String,
+        output_logs: OutputLogs,
+    },
+    TaskOutput {
+        task: String,
+        output: Vec<u8>,
+    },
+    EndTask {
+        task: String,
+        result: TaskResult,
+    },
+    Status {
+        task: String,
+        status: String,
+        result: CacheResult,
+    },
+    UpdateTasks {
+        tasks: Vec<String>,
+    },
+    Stop,
+}
+
+#[derive(Debug, Clone, Serialize)]
+pub struct ServerMessage<'a> {
+    pub id: u32,
+    #[serde(flatten)]
+    pub payload: &'a WebUIEvent,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(tag = "type", content = "payload")]
+pub enum ClientMessage {
+    /// Acknowledges the receipt of a message.
+    /// If we don't receive an ack, we will resend the message
+    Ack { id: u32 },
+    /// Asks for all messages from the given id onwards
+    CatchUp { start_id: u32 },
+}
+
+struct AppState {
+    rx: tokio::sync::broadcast::Receiver<WebUIEvent>,
+    // We use a tokio::sync::Mutex here because we want this future to be Send.
+    #[allow(clippy::type_complexity)]
+    messages: Arc<Mutex<RefCell<Vec<(WebUIEvent, u32)>>>>,
```

Comment:
I think we really want to use a new tokio task to manage `messages` state. If the socket is slow on receiving these, we can end up dropping a lot of messages due to how `tokio::broadcast` works.

We're doing a lot of sends on the socket handler task and if one of them ends up being slow we stop moving forward.